### PR TITLE
feat: Add MCTS agent

### PR DESCRIPTION
**Pull request by sorgfresser**
First of all, thank you very much for the project! It is of great use for me right now and I think it's an awesome contribution to Lean4 ATP.

This PR adds an interface for agents using monte carlo tree search for proof search.
I also added a dummy agent and some tests.
There will most likely be some things that are not ideal yet, I am more than willing to adjust them, just let me know! :)

One thing to note is `tactic_feedback` in Agent, which should most likely be implemented on `SearchState` for it to be useful.

### CODE REVIEW

**Comment by lenianiva** [comment]
Nice work! I'll merge it after we submit the main repository for TACAS'25 artefact

**Comment by lenianiva** [comment]
> One thing to note is tactic_feedback in Agent, which should most likely be implemented on SearchState for it to be useful.

I'll change this on main

**Comment by sorgfresser** [comment]
After reading the paper on Pantograph: would you want me to rename the selection step to `policy_function` to align it with the paper?

**Comment by lenianiva** [comment]
> After reading the paper on Pantograph: would you want me to rename the selection step to `policy_function` to align it with the paper?
> 
> Also, I am not 100% sure whether this implementation is treating goals as intended by the paper. My nodes are currently SearchStates, not goals. We choose goals using priorities, so the current policy function first chooses a proof state and then a goal using the proof state's priorities. This will definitely result in problems if we disable automatic mode.

The name `select` is fine. The policy function should be a mathematical function called in the select step.

>  My nodes are currently SearchStates, not goals.

This is perfect

> This will definitely result in problems if we disable automatic mode.

If the user disables automatic mode they're on their own.

**Comment by lenianiva** [comment]
> One thing to note is `tactic_feedback` in Agent, which should most likely be implemented on `SearchState` for it to be useful.

https://github.com/lenianiva/PyPantograph/pull/42/files#diff-5e097027bfb3ccab1ace21fb05276145f1b50947f1a81bb8beee74ff1082902a

Like this?

**Comment by sorgfresser** [comment]
> > One thing to note is `tactic_feedback` in Agent, which should most likely be implemented on `SearchState` for it to be useful.
> 
> https://github.com/lenianiva/PyPantograph/pull/42/files#diff-5e097027bfb3ccab1ace21fb05276145f1b50947f1a81bb8beee74ff1082902a
> 
> Like this?

That's also very helpful, but I was thinking about something else. I'll create a separate PR!

**Comment by lenianiva** [comment]
The two search branches are merged.

**Comment by sorgfresser** [comment]
Thanks for merging the other ones! 
I merged main into this for now, I can also rebase properly if you want to avoid merge commits.

**Comment by lenianiva** [comment]
> Thanks for merging the other ones! I merged main into this for now, I can also rebase properly if you want to avoid merge commits.

Merging is fine. You can merge main into your branch and then I'll merge your PR

**Comment by sorgfresser** [comment]
Should be done :)
Thanks!!

File: pantograph/search.py [hunk 1]
```diff
@@ -246,6 +470,31 @@ def test_solve_big(self):
             verbose=False)
         self.assertTrue(flag)
 
+class TestMCTSSearch(unittest.TestCase):
+
+    def test_solve(self):
```
Comment (lenianiva) [inline pantograph/search.py:475]: Could you modify these test cases so when they're successful they don't print anything to stdout? i.e. if you run `python3 -m pantograph.search` it should not output anything.

**Review by lenianiva — CHANGES_REQUESTED** [review]

File: pantograph/search.py (hunk 1 above)
Comment (sorgfresser) [reply pantograph/search.py:475]: Of course, sorry about that! I just changed the verbosity to False, this should fix it.

**Comment by lenianiva** [comment]
Approved.

**Review by lenianiva — APPROVED** [review]

**Pull request merged**